File: tests/board_tests/rowhammer_test.cpp
/**
 * Single-sided RowHammer test.
 *
 * For each victim row in [start_row, start_row + num_victims):
 *   1. Initialize victim row with DATA pattern
 *   2. Initialize aggressor row (victim + 1) with ~DATA pattern
 *   3. Hammer the aggressor row N times
 *   4. Read back the victim row
 *   5. Report bitflips
 */

#include <algorithm>
#include <charconv>
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <exception>
#include <span>
#include <string_view>
#include <vector>

#include "drambender/api/board/board.h"

using namespace DRAMBender;

namespace {

constexpr int k_bytes_per_cacheline = 64;
constexpr int k_cachelines_per_row = 128;
constexpr int k_row_bytes = k_bytes_per_cacheline * k_cachelines_per_row;
constexpr int k_words_per_cacheline = 16;
constexpr int k_column_stride = 8;

// Default config
constexpr int k_default_bank = 0;
constexpr int k_default_start_row = 0;
constexpr int k_default_num_victims = 64;
constexpr int k_default_hammer_count = 150000;
constexpr uint32_t k_victim_data = 0x00000000;
constexpr uint32_t k_aggressor_data = 0xFFFFFFFF;

// Register IDs
constexpr int CASR = 0;
constexpr int BASR = 1;
constexpr int RASR = 2;
constexpr int CAR = 3;
constexpr int BAR = 4;
constexpr int RAR = 5;
constexpr int PATTERN = 6;
constexpr int NUM_HAMMER = 7;
constexpr int HAMMER_CTR = 8;

struct Options {
  int instance_id = 0;
  int bank = k_default_bank;
  int start_row = k_default_start_row;
  int num_victims = k_default_num_victims;
  int hammer_count = k_default_hammer_count;
};

bool parse_int(const char* text, int* value) {
  std::string_view input(text);
  int parsed = 0;
  auto result = std::from_chars(input.data(), input.data() + input.size(), parsed);
  if (result.ec != std::errc() || result.ptr != input.data() + input.size()) return false;
  *value = parsed;
  return true;
}

bool parse_args(int argc, char** argv, Options* opts) {
  for (int i = 1; i < argc; i += 2) {
    std::string_view arg(argv[i]);
    if (i + 1 >= argc) { std::fprintf(stderr, "Missing value for %.*s\n", (int)arg.size(), arg.data()); return false; }
    int* target = nullptr;
    if (arg == "--instance-id") target = &opts->instance_id;
    else if (arg == "--bank") target = &opts->bank;
    else if (arg == "--start-row") target = &opts->start_row;
    else if (arg == "--num-victims") target = &opts->num_victims;
    else if (arg == "--hammer-count") target = &opts->hammer_count;
    else { std::fprintf(stderr, "Unknown argument: %.*s\n", (int)arg.size(), arg.data()); return false; }
    if (!parse_int(argv[i + 1], target)) { std::fprintf(stderr, "Invalid value for %.*s\n", (int)arg.size(), arg.data()); return false; }
  }
  return true;
}

// Build a program that: init victim, init aggressor, hammer, read victim
FinalProgram build_rowhammer_program(int bank, int victim_row, int aggressor_row,
                                     uint32_t victim_pattern, uint32_t aggressor_pattern,
                                     int hammer_count) {
  Program p;

  p.add_inst(SMC_LI(bank, BAR));
  p.add_inst(SMC_LI(k_column_stride, CASR));
  p.add_inst(SMC_LI(hammer_count, NUM_HAMMER));
  p.add_inst(SMC_LI(0, HAMMER_CTR));

  // --- Initialize victim row ---
  p.add_inst(SMC_LI(victim_row, RAR));
  p.add_inst(SMC_LI(victim_pattern, PATTERN));
  for (int w = 0; w < k_words_per_cacheline; ++w)
    p.add_inst(SMC_LDWD(PATTERN, w));

  p.add_inst(SMC_PRE(BAR, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(SMC_LI(0, CAR));
  p.add_inst(all_nops());
  p.add_inst(all_nops());
  p.add_inst(SMC_ACT(BAR, 0, RAR, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(all_nops());
  p.add_inst(all_nops());
  for (int i = 0; i < k_cachelines_per_row; ++i) {
    p.add_inst(SMC_WRITE(BAR, 0, CAR, 1, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
    p.add_inst(all_nops());
  }
  p.add_inst(SMC_SLEEP(8));
  p.add_inst(SMC_PRE(BAR, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(all_nops());
  p.add_inst(all_nops());

  // --- Initialize aggressor row ---
  p.add_inst(SMC_LI(aggressor_row, RAR));
  p.add_inst(SMC_LI(aggressor_pattern, PATTERN));
  for (int w = 0; w < k_words_per_cacheline; ++w)
    p.add_inst(SMC_LDWD(PATTERN, w));

  p.add_inst(SMC_PRE(BAR, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(SMC_LI(0, CAR));
  p.add_inst(all_nops());
  p.add_inst(all_nops());
  p.add_inst(SMC_ACT(BAR, 0, RAR, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(all_nops());
  p.add_inst(all_nops());
  for (int i = 0; i < k_cachelines_per_row; ++i) {
    p.add_inst(SMC_WRITE(BAR, 0, CAR, 1, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
    p.add_inst(all_nops());
  }
  p.add_inst(SMC_SLEEP(8));
  p.add_inst(SMC_PRE(BAR, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(all_nops());
  p.add_inst(all_nops());

  // --- Hammer aggressor ---
  // Loop body: PRE → (increment) → ACT → (branch back to PRE)
  // This ensures ACT is always followed by PRE on the next iteration.
  p.add_inst(SMC_LI(aggressor_row, RAR));
  p.add_inst(SMC_LI(0, HAMMER_CTR));

  // Prime: first ACT before entering the loop (tRAS satisfied by loop body)
  p.add_inst(SMC_ACT(BAR, 0, RAR, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(all_nops());
  p.add_inst(all_nops());
  p.add_inst(all_nops());
  p.add_inst(all_nops());
  p.add_inst(all_nops());

  p.add_label("HAMMER");
  // PRE (row was activated above or at bottom of loop)
  p.add_inst(SMC_PRE(BAR, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(SMC_ADDI(HAMMER_CTR, 1, HAMMER_CTR));
  p.add_inst(all_nops());
  // ACT at end of word — next iteration starts with PRE
  p.add_inst(SMC_NOP(), SMC_NOP(), SMC_NOP(), SMC_ACT(BAR, 0, RAR, 0));
  p.add_branch(Program::BR_TYPE::BL, HAMMER_CTR, NUM_HAMMER, "HAMMER");

  // Final PRE after loop exits (last ACT needs closing)
  p.add_inst(SMC_PRE(BAR, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(all_nops());
  p.add_inst(all_nops());

  // --- Read victim row back ---
  p.add_inst(SMC_LI(victim_row, RAR));
  p.add_inst(SMC_LI(0, CAR));
  p.add_inst(SMC_PRE(BAR, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(all_nops());
  p.add_inst(all_nops());
  p.add_inst(SMC_ACT(BAR, 0, RAR, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(all_nops());
  p.add_inst(all_nops());
  for (int i = 0; i < k_cachelines_per_row; ++i) {
    p.add_inst(SMC_READ(BAR, 0, CAR, 1, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
    p.add_inst(all_nops());
  }
  p.add_inst(SMC_SLEEP(4));
  p.add_inst(SMC_PRE(BAR, 0, 0), SMC_NOP(), SMC_NOP(), SMC_NOP());
  p.add_inst(all_nops());
  p.add_inst(all_nops());

  return p.conclude();
}

size_t count_bitflips(std::span<const std::byte> row_buffer, uint32_t expected_pattern) {
  size_t flips = 0;
  const auto* words = reinterpret_cast<const uint32_t*>(row_buffer.data());
  size_t num_words = row_buffer.size() / sizeof(uint32_t);
  for (size_t i = 0; i < num_words; ++i) {
    uint32_t xored = words[i] ^ expected_pattern;
    flips += __builtin_popcount(xored);
  }
  return flips;
}

}  // namespace

int main(int argc, char** argv) {
  Options opts;
  if (!parse_args(argc, argv, &opts)) return 2;

  try {
    auto board = create_board(BoardType::DDR4, opts.instance_id, HostInterface::XDMA);
    board->reset_fpga();

    std::printf("rowhammer_test: bank=%d start_row=%d num_victims=%d hammer_count=%d\n",
                opts.bank, opts.start_row, opts.num_victims, opts.hammer_count);

    std::vector<std::byte> row_buffer(k_row_bytes);
    size_t total_flips = 0;
    size_t vulnerable_rows = 0;

    for (int v = 0; v < opts.num_victims; ++v) {
      int victim_row = opts.start_row + v;
      int aggressor_row = victim_row + 1;

      auto program = build_rowhammer_program(
          opts.bank, victim_row, aggressor_row,
          k_victim_data, k_aggressor_data, opts.hammer_count);
      board->execute(program);
      board->receive(std::span(row_buffer));
      board->synchronize();

      size_t flips = count_bitflips(row_buffer, k_victim_data);
      total_flips += flips;
      if (flips > 0) {
        ++vulnerable_rows;
        std::printf("  row %5d: %zu bitflips (aggressor=%d)\n",
                    victim_row, flips, aggressor_row);
      }

      if ((v + 1) % 10 == 0 || v + 1 == opts.num_victims) {
        std::printf("\r  [%d/%d victims tested, %zu vulnerable, %zu total bitflips]",
                    v + 1, opts.num_victims, vulnerable_rows, total_flips);
        std::fflush(stdout);
      }
    }

    std::printf("\n\nResult: %zu/%d rows vulnerable, %zu total bitflips\n",
                vulnerable_rows, opts.num_victims, total_flips);
    return vulnerable_rows > 0 ? 1 : 0;
  } catch (const std::exception& e) {
    std::fprintf(stderr, "Error: %s\n", e.what());
    return 2;
  }
}

File: tests/board_tests/rowhammer_test.py
#!/usr/bin/env python3
"""Single-sided RowHammer test.

For each victim row in [start_row, start_row + num_victims):
  1. Initialize victim row with DATA pattern
  2. Initialize aggressor row (victim + 1) with ~DATA pattern
  3. Hammer the aggressor row N times
  4. Read back the victim row
  5. Report bitflips
"""

import argparse
import sys
from functools import partial
from pathlib import Path

_REPO_ROOT = Path(__file__).resolve().parents[2]
if str(_REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(_REPO_ROOT))

import numpy as np  # noqa: E402

import drambender  # noqa: E402
from tests.jit_benchmark.workloads import (  # noqa: E402
    CACHELINES_PER_ROW,
    ROW_BYTES,
    build_rowhammer_program_compiled,
)

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

WORDS_PER_ROW = ROW_BYTES // 4


# ---------------------------------------------------------------------------
# Bitflip analysis (vectorized)
# ---------------------------------------------------------------------------

def get_bitflip_indices(observed_bytes: np.ndarray, expected_pattern: int) -> np.ndarray:
    """Return global bit indices of all flipped bits."""
    observed_u32 = observed_bytes.view(np.uint32)
    expected_u32 = np.full(len(observed_u32), expected_pattern, dtype=np.uint32)
    flips = observed_u32 ^ expected_u32
    bits = np.unpackbits(flips.view(np.uint8), bitorder="little")
    return np.flatnonzero(bits).astype(np.int64)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> int:
    ap = argparse.ArgumentParser(description="Single-sided RowHammer test")
    ap.add_argument("--board-id", type=int, default=0)
    ap.add_argument("--instance-id", type=int, default=0)
    ap.add_argument("--bank", type=int, default=0)
    ap.add_argument("--start-row", type=int, default=0)
    ap.add_argument("--num-victims", type=int, default=64)
    ap.add_argument("--hammer-count", type=int, default=150000)
    ap.add_argument("--victim-data", type=lambda x: int(x, 0), default=0x00000000)
    ap.add_argument("--aggressor-data", type=lambda x: int(x, 0), default=0xFFFFFFFF)
    args = ap.parse_args()

    print(f"rowhammer_test: board={args.board_id} instance={args.instance_id} "
          f"bank={args.bank} start_row={args.start_row} "
          f"num_victims={args.num_victims} hammer_count={args.hammer_count}")
    print(f"  victim_data=0x{args.victim_data:08x} aggressor_data=0x{args.aggressor_data:08x}")

    board = drambender.api.DDR4(args.board_id, args.instance_id)
    board.reset_fpga()

    # Build template once, bind fixed args
    hammer = partial(
        build_rowhammer_program_compiled,
        bank=args.bank,
        victim_pattern=args.victim_data,
        aggressor_pattern=args.aggressor_data,
        hammer_count=args.hammer_count,
    )

    row_buffer = np.empty(ROW_BYTES, dtype=np.uint8)
    total_flips = 0
    vulnerable_rows = 0

    for v in range(args.num_victims):
        victim_row = args.start_row + v
        aggressor_row = victim_row + 1

        program = hammer(victim_row=victim_row, aggressor_row=aggressor_row)
        board.execute(program)
        board.receive_into(row_buffer)
        board.synchronize()

        flip_indices = get_bitflip_indices(row_buffer, args.victim_data)
        n_flips = len(flip_indices)
        total_flips += n_flips

        if n_flips > 0:
            vulnerable_rows += 1
            print(f"  row {victim_row:5d}: {n_flips} bitflips (aggressor={aggressor_row})")
            for bit_idx in flip_indices[:16]:
                word_idx = int(bit_idx) // 32
                bit_in_word = int(bit_idx) % 32
                cl = word_idx // 16
                print(f"    cl={cl} word={word_idx % 16} bit={bit_in_word}")

        done = v + 1
        if done % 10 == 0 or done == args.num_victims:
            print(f"\r  [{done}/{args.num_victims} tested, "
                  f"{vulnerable_rows} vulnerable, {total_flips} total bitflips]",
                  end="", flush=True)

    print(f"\n\nResult: {vulnerable_rows}/{args.num_victims} rows vulnerable, "
          f"{total_flips} total bitflips")
    return 1 if vulnerable_rows > 0 else 0


if __name__ == "__main__":
    sys.exit(main())
